### DISCUSSION TITLE
Fine-grained Personal Access Token permissions

### DISCUSSION THREAD

**Discussion by jbecker-idt** [community/community#133558] [New to GitHub] [answered]
Hello,

I was wondering if it is necessary to use personal access tokens in order to push to a repository, and if it is, is there is a more approachable explanation of the what the permissions under both the Fine-grained and Classic Personal Access Tokens means other than the one github has published? and if it's not, what way would you guys would recommend, and why you would recommend executing pushes that way.

I've been making my way through the ProGit text book written by Scott Chacon and Ben Straub. The book instructs students to familiarize themselves with git through command line. I've been using git bash. I was having trouble executing pushes to my practice repository until I enabled several options under the Fine-grained Personal Access Token permissions settings; the trouble is, I'm not finding the description of the permission settings to be organized or written for a beginner audience-- I'm still not entirely sure which permissions are bare the minimum to execute a push, or what vulnerabilities they create. It leaves me wondering if there is another way beginners should be interacting with their repositories until they have the git workflow under their fingertips.

**Comment by unknownguy49** [comment, marked as answer]
Yes, using personal access tokens (PATs) is necessary for pushing to GitHub repositories, as GitHub has deprecated password authentication for Git operations. Here's a simpler explanation of the permissions:

### Fine-grained Personal Access Tokens
1. **Repository permissions:**
   - **Contents:** Access to repository contents.
   - **Metadata:** Access to repository metadata.
   - **Pull requests:** Access to pull request information.
   - **Issues:** Access to issue information.

   For basic push operations, you need:
   - **Contents:** Read and write access.
   - **Metadata:** Read access.

### Classic Personal Access Tokens
1. **repo:** Full control of private repositories.
   - This includes read/write access to code, metadata, pull requests, and issues.

### Recommendations for Beginners
1. **Fine-grained PATs:** Use with minimal permissions (e.g., Contents: read/write, Metadata: read).
2. **Classic PATs:** Easier for initial setup (use the `repo` scope).
3. **SSH Keys:** More secure and easier in the long run, especially if you're using Git Bash.

### Setting up SSH Keys
1. Generate an SSH key:
   ```bash
   ssh-keygen -t ed25519 -C "your_email@example.com"
   ```
2. Add the SSH key to the ssh-agent:
   ```bash
   eval "$(ssh-agent -s)"
   ssh-add ~/.ssh/id_ed25519
   ```
3. Add the SSH key to your GitHub account (found in settings under SSH and GPG keys).
4. Clone/push using SSH:
   ```bash
   git clone git@github.com:username/repo.git
   git push origin main
   ```

Using SSH keys is recommended as it avoids the need to manage tokens and provides a secure connection.
